test/firtool/sink-probes-into-layers.fir

```diff
@@ -0,0 +1,40 @@
+; RUN: firtool --disable-all-randomization --advanced-layer-sink %s | FileCheck %s
+
+; Check that the probes are correctly sunk into the layer, and there are
+; no bare probes left in the design.
+
+; Check the probes are inlined into the assertions:
+
+; CHECK-LABEL: module Foo_A();
+; CHECK-NEXT:    always @(posedge Foo.clock) begin
+; CHECK-NEXT:      assert(Foo.bar.`ref_Bar_probe) else $error("message");
+; CHECK-NEXT:      assert(Foo.bar.`ref_Bar_probe) else $error("message");
+; CHECK-NEXT:    end // always @(posedge)
+; CHECK-NEXT:  endmodule
+
+; Check that there are no probes in the top level module:
+
+; CHECK-LABEL: module Foo(
+; CHECK-NEXT:    input clock
+; CHECK-NEXT:  );
+; CHECK-EMPTY:
+; CHECK-NEXT:    Bar bar ();
+; CHECK-NEXT:  endmodule
+
+FIRRTL version 4.0.0
+
+circuit Foo:
+  layer A, bind:
+
+  public module Foo:
+    input clock: Clock
+
+    inst bar of Bar
+    node cond = read(bar.probe)
```

Comment:
Thanks for the great end-to-end test, especially given the pipeline-sensitive nature of the interplay between layers and probes (mostly historical, but we may yet want to shuffle them a little).

Reply:
Maybe including a rwprobe (+force inside the layerblock) to ensure the rwprobe is sunk too? It's not the same since it's not "dereferenced" outside the layer like the bare `read(..)` here, so up to you.